### DISCUSSION TITLE
statically Inbuilt legacy module and dynamic fips module loading issue

### DISCUSSION THREAD

**Discussion by aniljadaun** [openssl/openssl#21444] [Q&A] [answered]
Hi

I am trying to create a legacy module inbuilt static opnessl libs, i used no-module no-shared option for configuration. Generated openssl libs succeeded in loading legacy module but doing so my fips module is not loading any more . 

Is it not possible do both at same time as no-module means no separate module and fips module is a dynamic module which loaded after self test ? 
We would prefer legacy to be inbuilt within openssl libs, though having fips compliance is mandatory for us. Any way to achieve both together ? 
Am i missing something here ?

**Comment by aniljadaun** [comment]
@mattcaswell @t8m can you please review the above query ?

**Comment by mattcaswell** [comment]
It should work fine but you have to build OpenSSL twice.

The first time you build OpenSSL with "no-module no-shared". This should give you static libraries with the legacy provider built-in.

The second time you build openssl with "enable-fips". Do not specify "no-module" or "no-shared". The only file you are interested in from this build is the fips.dll/fips.so file. Install it as per the FIPS module security policy into the MODULESDIR as appropriate for your first build.

**Reply by paulidale** [reply]
Also note that the second build doesn't have to be the same version of OpenSSL as the first.  Specifically, OpenSSL 3.0.0 and 3.0.8 are the only currently FIPS validated versions.  You must use one of these two to get a FIPS provider than is compliant.

We generally suggest using versions 3.0.9 or 3.1.1 with the 3.0.8 FIPS provider.

**Reply by aniljadaun** [reply]
Still not working...This could be a bug as well for the combination legacy inbuilt static libs with fips module. Can you please try this at your end.
This works for me using openssl libs without inbuilt legacy module

Path for MODULEDIR is still coming as the path of legacy libs generated path for ossl-modules
Even copying fips.dll to this path didn't work.


Sample Code:
{
	const char* default_search_path = "D:\\fipsdll\\";
	OSSL_PROVIDER_set_default_search_path(NULL, default_search_path);
	const char* modulesdir = OpenSSL_version(OPENSSL_MODULES_DIR);

// updated openssl_modules environment variable
	std::string envVar = "OPENSSL_MODULES=";
	envVar += default_search_path;
	int result = _putenv(const_cast<char*>(envVar.c_str()));
	if (result) {
		cout << "Environment variable not set" << endl;
	}
	if (!OSSL_LIB_CTX_load_config(NULL, "D:\\outlibsopenssl\\SSL\\openssl.cnf")) {
                 cout<<"conf file not loaded"<<endl;
		}
	OSSL_PROVIDER*  fips; = OSSL_PROVIDER_load(NULL, "fips");
	modulesdir = OpenSSL_version(OPENSSL_MODULES_DIR);
	if (fips == NULL) {
		cout<<"fips moduel not loaded"<<endl;
		}
		// exit(EXIT_FAILURE);
	}
	

Openssl.cnf file
------------------------------------------------
config_diagnostics = 1
openssl_conf = openssl_init

.include "D:\\outlibsopenssl\\SSL\\fipsmodule.cnf"

[openssl_init]
providers = provider_sect


[provider_sect]
fips = fips_sect
base = base_sect

[base_sect]
activate = 1

fipsmodule.cnf file
------------------------------------------------
[fips_sect]
activate = 1
install-version = 1
conditional-errors = 1
security-checks = 1
module-mac = FB:C6:40:EF:3F:61:01:29:35:E9:AC:9A:B4:0A:1E:E8:AB:94:08:B0:5B:B0:AC:50:E3:27:05:C4:ED:A1:31:33
install-mac = 41:9C:38:C2:8F:59:09:43:2C:AA:2F:58:36:2D:D9:04:F9:6C:56:8B:09:E0:18:3A:2E:D6:CC:69:05:04:E1:11
install-status = INSTALL_SELF_TEST_KATS_RUN

Error i get is : 
F0190000:error:07880025:common libcrypto routines:provider_init:reason(37):crypto\provider_core.c:914:name=fips

@mattcaswell @t8m @paulidale  please suggest

**Reply by t8m** [reply]
I've tested (with openssl-3.0.9 on Linux) that I am able to do `no-module` build and use a fips provider from another build with it.

**Reply by t8m** [reply]
There was a mistake in my testing - I was using the libcrypto.so.3 from the system and not the one I've built with no-module.

**Comment by t8m** [comment]
> Error i get is : F0190000:error:07880025:common libcrypto routines:provider_init:reason(37):crypto\provider_core.c:914:name=fips
> 
> @mattcaswell @t8m @paulidale please suggest

Is this the only error on the stack? Have you printed all the errors? https://www.openssl.org/docs/man3.0/man3/ERR_print_errors_fp.html

**Comment by t8m** [comment, marked as answer]
@aniljadaun When building the openssl in the no-module configuration, do you see whether the sources from crypto/dso are built?

~~For some reason they are being built on Linux even though no-module implies no-dso (which is actually incorrect).~~

Could you try this patch:

```diff
diff --git a/Configure b/Configure
index 4ddc275f78..099568e679 100755
--- a/Configure
+++ b/Configure
@@ -616,7 +616,7 @@ my @disable_cascades = (
     # or modules.
     "pic"               => [ "shared", "module" ],
 
-    "module"            => [ "fips", "dso" ],
+    "module"            => [ "fips" ],
 
     "engine"            => [ "dynamic-engine", grep(/eng$/, @disablables) ],
     "dynamic-engine"    => [ "loadereng" ],
```

**Reply by aniljadaun** [reply]
@t8m Thanks for your suggestion. I will try this. Just to let you know i am doing no-shared no-module for windows.

**Reply by aniljadaun** [reply]
@t8m This worked. Thanks alot for your feedback .:)

**Comment by t8m** [comment]
So this was a real bug - see https://github.com/openssl/openssl/pull/21459

**Reply by aniljadaun** [reply]
Hi @mattcaswell @t8m 

I am using dynamic libs on Mac. Is it possible to create inbuilt legacy module in the dynamic libs , also we would openssl to be fips compliant. please suggest

perl Configure no-module should work ? any steps other then this ?

**Reply by t8m** [reply]
Hello,  perl Configure no-module was supposed to work but unfortunately due to bug #21459 it does not. It should be fixed in the next release.

**Comment by ChauhanPriyanka** [comment]
Hi @t8m , in which OpenSSL 3.0 version the fix for the issue will be available, is it already released. I'm not able to load fips.dll with inbuild legacy provider. I'm testing with openssl 3.0.9

**Reply by mattcaswell** [reply]
This was fixed in 3.0.10 and 3.1.2

**Reply by ChauhanPriyanka** [reply]
Thank you @mattcaswell
